Update WS testing coding standards doc (#12213)
bbc/simorgh: docs/Test-Strategy-Info.mdx

import { Meta } from '@storybook/blocks';
import { Link } from 'react-router-dom';

<Meta title="docs/WS-Automation-strategy" />

# World Service - Test Automation strategy

## Background 

This document aims to provide high-level guidance on approaching test automation for World Service teams. It specifically seeks to answer the following questions:

- Why should we automate? (Why we write tests)
- Where should we automate? (What levels)
- What should we automate? 
- When should we automate?
- How should we automate? (Patterns, Anti-Patterns and Practices)

> - Who should automate? (Responsibilities)

Reviewer: Should we consider linking to parts further down in the document, or perhaps a table of contents?

Author: I've added table of contents.

#### Critical paths 
The critical path is the set of components and workflows that is required for the application to serve its core function. For example, an article page displaying the content and title for the given service is critical in our context, but links to related topics or top stories, can be considered non-critical. These should be tested as well, but they’re better candidates for a less expensive layer of testing like integration tests.

#### Smoke tests
We should have smoke tests for validating that core functionality is not broken, for example, that the page loads, has titles and content. These should run as often as possible (on each PR), they should be quick, reliable and repeatable (not flaky).

#### As little as possible
The rule of thumb is that we should avoid writing an end to end test if there is an alternative to validating the functionality. The alternative can be a combination of layers, for example, an integration test combined with an API test (which doesn’t exist right now) can be enough to reach the same confidence as an e2e test with less cost.

> ### Puma approach

Reviewer: Use caps since this is an acronym
```suggestion
### PUMA approach
```

> Adopting the [PUMA](https://confluence.dev.bbc.co.uk/display/podtest/PUMA) approach can provide a framework to decide if a proposed test genuinely meets the definition of an E2E. According to PUMA, an E2E test should:

Reviewer: This is an internal confluence link - can we link to a public source? Or is it OK to do this - let's get a second opinion from @andrewscfc

Reviewer: I would like to use a link from the web rather than an internal confluence page if possible - can we try to find one?

> P - Prove core functionality

Reviewer: There needs to be a whitespace here in order to get P... on a new line
```suggestion
Adopting the [PUMA](https://confluence.dev.bbc.co.uk/display/podtest/PUMA) approach can provide a framework to decide if a proposed test genuinely meets the definition of an E2E. According to PUMA, an E2E test should:

P - Prove core functionality
```

> 
> If the test is “green” (i.e. it confirms the actual behaviour matches the expected behaviour), then we are provided with the knowledge that at that time the system behaved as defined by the test.
> 
> If the test is “red” (i.e. it confirms the actual behaviour did not match expectation), then we are provided with the knowledge that at that time the system did not behave as defined by the test.
> 
> If a test remains permanently “red”, then we either misunderstood the purpose of the code and so need to review the expectations of the checks in the test or we have uncovered unexpected behaviour in the system under test.
> 
> If a test remains permanently “green”, then we confirm current behaviour continues to meet the expectations of the test.
> 
> However, a test that flickers between the two states indicates something is not right and should not be ignored! I believe describing these as flaky makes it easy to ignore potential problems either in the functionality of the system under test or the test itself.
> 
> quote from  @sframpton

> > 🔊 Discussion with team: What are other good practices? any anti-patterns?

Reviewer: ```suggestion
> 🔊 Discussion with team: 
> What are other good practices? 
> Are there any anti-patterns?
```

import { Meta } from '@storybook/blocks';
import { Link } from 'react-router-dom';

<Meta title="docs/WS-Automation-strategy" />

> # WS - Automation strategy

Reviewer: ```suggestion
# World Service - Test Automation strategy
```

## Background 

This document aims to provide high-level guidance on approaching test automation for World Service teams. It specifically seeks to answer the following questions:

- Why should we automate? (Why we write tests)
- Where should we automate? (What levels)
- What should we automate? 
- When should we automate?
- How should we automate? (Patterns, Anti-Patterns and Practices)
- Who should automate? (Responsibilities)

> While this document focuses primarily on e2e test strategy, it touches on other levels of testing to give a holistic understanding and context for the decisions we make when it comes to testing. It documents some of our existing approaches, while suggesting gradual improvements for the future. It’s a living collaborative document that we should refer to and update continuously. 

Reviewer: Let's use consistent case for E2Es across the document

```suggestion
While this document focuses primarily on E2E test strategy, it touches on other levels of testing to give a holistic understanding and context for the decisions we make when it comes to testing. It documents some of our existing approaches, while suggesting gradual improvements for the future. It’s a living collaborative document that we should refer to and update continuously. 
```

> > You can also check this [related document](https://onebbc-my.sharepoint.com/:w:/g/personal/meriem_mechri01_bbc_co_uk/EftDa1m66I5FuNGzD2gDEdYBSoBe-g9LDAvJf91_K6JuLw?email=Simon.Frampton%40bbc.co.uk&e=jkNFze&wdOrigin=TEAMS-MAGLEV.p2p_ns.rwc&wdExp=TEAMS-TREATMENT&wdhostclicktime=1722460818407&web=1) that describes our current e2e stack.

Reviewer: Should we be referencing a personal MS Doc in a public readme? Perhaps it's best to add its contents to a readme here too.

Author: I have removed this link, until I find a way to move it public.

> > You can also check this [related document](https://onebbc-my.sharepoint.com/:w:/g/personal/meriem_mechri01_bbc_co_uk/EftDa1m66I5FuNGzD2gDEdYBSoBe-g9LDAvJf91_K6JuLw?email=Simon.Frampton%40bbc.co.uk&e=jkNFze&wdOrigin=TEAMS-MAGLEV.p2p_ns.rwc&wdExp=TEAMS-TREATMENT&wdhostclicktime=1722460818407&web=1) that describes our current e2e stack.

Reviewer: ```suggestion
> You can also check this [related document](https://onebbc-my.sharepoint.com/:w:/g/personal/meriem_mechri01_bbc_co_uk/EftDa1m66I5FuNGzD2gDEdYBSoBe-g9LDAvJf91_K6JuLw?email=Simon.Frampton%40bbc.co.uk&e=jkNFze&wdOrigin=TEAMS-MAGLEV.p2p_ns.rwc&wdExp=TEAMS-TREATMENT&wdhostclicktime=1722460818407&web=1) that describes our current E2E stack.
```

## Why should we automate?

There are many reasons why we write automated tests. Some of the most common are:

1. Verify the code is working correctly
2. Prevent future regressions
3. Document code behaviour
4. Provides (code) design guidance
5. Support refactoring

[source: https://madeintandem.com/blog/five-factor-testing/]

> > 🔊 discussion point with team: any other reasons? what are for you the most important reasons?

Reviewer: ```suggestion
> 🔊 Discussion point with team:
> - Any other reasons? 
> - What are for you the most important reasons?
```

> It is important to think about why  we write tests because it influences what types of testing we should prioritise. For example, unit tests might be good for improving the system design as the units become more independent and easier to test, but having only unit tests doesn’t necessarily verify the code is working correctly (beyond the single unit) as it won’t necessarily reveal problems that occur in the integration of components. On the other hand, E2E tests can provide better verification of the software, but they’re expensive and would not necessarily improve software design, nor document the code behaviour at unit or component level.

Reviewer: ```suggestion
It is important to think about why we write tests because it influences what types of testing we should prioritise. For example, unit tests might be good for improving the system design as the units become more independent and easier to test, but having only unit tests doesn’t necessarily verify the code is working correctly (beyond the single unit) as it won’t necessarily reveal problems that occur in the integration of components. On the other hand, E2E tests can provide better verification of the software, but they’re expensive and would not necessarily improve software design, nor document the code behaviour at unit or component level.
```

## Where should we automate?

> ## Defining the test levels

Reviewer: I think we might need to change the heading levels for GH
```suggestion
## Where should we automate?
### Defining the test levels
```

> Broadly speaking, we follow the test pyramid. We have a large set of unit tests, a smaller subset of integration tests, and an even smaller subset of UI tests. We should continue to follow this pattern.

Reviewer: Do we need to be clear what UI tests are - or should this be E2E?

```suggestion
Broadly speaking, we follow the test pyramid. We have a large set of unit tests, a smaller subset of integration tests, and an even smaller subset of UI tests. We should continue to follow this pattern.
```

![Image](https://blog.getmason.io/content/images/2020/11/Testing-pyramid--6--1.jpg)
Image [Source](https://blog.getmason.io/content/images/2020/11/Testing-pyramid--6--1.jpg)
To avoid confusion, we will clarify what the test levels mean in our context.

> ### Unit tests

Reviewer: We're going to look at all headings in this doc to decide the correct level. Let's do this at the end, once all other amendments have been made/

We use jest to write unit tests
They normally exercise a single component (UI components, hooks, helper methods)

> Some examples of our unit tests can be found [here](https://github.com/bbc/simorgh/tree/bfbed54d2b1fe880e8f171023054c16d1c274ca2/src/app/components/Heading).

Reviewer: ```suggestion
We use jest to write unit tests. They normally exercise a single component (UI components, hooks, helper methods). Here is an [example unit test](https://github.com/bbc/simorgh/tree/bfbed54d2b1fe880e8f171023054c16d1c274ca2/src/app/components/Heading).
```

Some patterns:

> - We use data-driven tests sometimes like [here](https://github.com/bbc/simorgh/blob/bfbed54d2b1fe880e8f171023054c16d1c274ca2/src/app/hooks/useImageColour/index.test.js#L10) to define inputs to tests

Reviewer: ```suggestion
- We use [data-driven tests](https://github.com/bbc/simorgh/blob/bfbed54d2b1fe880e8f171023054c16d1c274ca2/src/app/hooks/useImageColour/index.test.js#L10) to define inputs to tests
```

> - We use snapshot tests to easily test some cases like [here](https://github.com/bbc/simorgh/blob/4f385d94bee96af48eff2eca49c24cf382a3a494/src/app/components/FrostedGlassPromo/index.test.tsx#L76)

Reviewer: ```suggestion
- We use [snapshot tests](https://github.com/bbc/simorgh/blob/4f385d94bee96af48eff2eca49c24cf382a3a494/src/app/components/FrostedGlassPromo/index.test.tsx#L76) to easily test certain scenarios such as (TODO what does this test show us?)
```

    

### Component tests
We conduct visual regression testing using Chromatic, which alongside our unit tests, comprise our component tests layer.

### Integration tests
These test the integration between the different components to ensure a page is rendered properly. 

> They typically check that what’s rendered matches SIMORGH_DATA in a flexible generic way, i.e. check that a page renders “most read” topics but not the exact topics rendered, and that each topic has an image and header but not the content of either.

Reviewer: ```suggestion
They typically check that what’s rendered matches local fixture data in a flexible generic way, i.e. check that a page renders “most read” articles but not the exact items rendered, and that each topic has an image and header but not the content of either.
```

Reviewer: Should we link to the code examples for these examples e.g. where we check that a page has most read or where we check the topic image & header

> Examples can be found in  [src/integration](https://github.com/bbc/simorgh/tree/bfbed54d2b1fe880e8f171023054c16d1c274ca2/src/integration).

Reviewer: ```suggestion
Examples can be found in [src/integration](https://github.com/bbc/simorgh/tree/bfbed54d2b1fe880e8f171023054c16d1c274ca2/src/integration).
```

> Check [here](https://github.com/bbc/simorgh/tree/latest/src/integration) for more details about our integration tests.

Reviewer: This technically links to the same directory as the line above. 

Should we instead link to the Inegration Testing docs? https://github.com/bbc/simorgh/blob/cbf7c2b3d08a33775c51eeb94de750d0993a988b/src/integration/README.mdx#integration

> ### UI (E2E) tests

Reviewer: Can we use consistent terminology. The engineers don't refer to the Cypress tests as UI tests. We would tend to refer to the Chromatic tests as UI tests... Shall we remove UI and just keep E2E?

> These tests run the actual UI in a browser, and can check that everything renders correctly. For example, an integration test might check that “most read” topics are rendered with the correct HTML tags, but that does not necessarily mean they appear correctly (some CSS value might hide them). With the UI tests, we check that everything shows as expected in the UI.

Reviewer: I am not quite sure how to reword this, but I think it should be something along the lines of 

These tests run against the application and 

```suggestion
These tests run against the application and can check that everything renders correctly. For example, an integration test might check that “most read” topics are rendered with the correct HTML tags, but that does not necessarily mean they appear correctly (some CSS value might hide them). With the UI tests, we check that everything shows as expected in the UI.
```

> These tests run the actual UI in a browser, and can check that everything renders correctly. For example, an integration test might check that “most read” topics are rendered with the correct HTML tags, but that does not necessarily mean they appear correctly (some CSS value might hide them). With the UI tests, we check that everything shows as expected in the UI.

Reviewer: ```suggestion
These tests run the actual UI in a browser, and can check that everything renders correctly. For example, an integration test might check that “most read” articles are rendered with the correct HTML tags, but that does not necessarily mean they appear correctly (some CSS value might hide them). With the E2E tests, we check that everything shows as expected in the UI.
```

Examples and instructions can be found in [cypress](https://github.com/bbc/simorgh/tree/latest/cypress).

> > 🔊 discussion point with team: should we break UI / e2e tests as @Simon F comment mentions - I agree with the comment, but I am not sure this reflects what we do in our team right now?

Reviewer: I think we should remove this
```suggestion
```

#### What levels to prioritise?

> Unit tests responsibility and scope are relatively easy to define: we already have a high level of code coverage and developers should aim to maintain that. 

Reviewer: ```suggestion
The responsibility and scope of unit tests are relatively easy to define: we already have a high level of code coverage and developers should aim to maintain that. 
```

> For integration and e2e tests, the general rule of thumb should be to only add an e2e test when an integration test is not enough to ensure coverage. Integration tests are less expensive and more stable and they should be prioritised.

Reviewer: Should we give an example of when an integration test is not enough to ensure coverage - usually this is some form of interactivity, navigtion or running asynchronous scripts on the page. 

Perhaps we could also link to the section of the integration docs which mention this - there's a bit here https://github.com/bbc/simorgh/blob/latest/src/integration/README.mdx#i-already-wrote-unit-tests-for-a-component-do-i-need-to-write-an-integration-test-for-this-component but not sure if that's enough

```suggestion
For integration and E2E tests, the general rule of thumb should be to only add an E2E test when an integration test is not enough to ensure coverage. Integration tests are less expensive and more stable and they should be prioritised.
```

> > ℹ️ Think about the maintenance cost of adding a new e2e test. Always favour an integration test if possible instead of an end to end test.

Reviewer: ```suggestion
> [!NOTE]
> Think about the maintenance cost of adding a new E2E test. Always favour an integration test if possible.
```

Reviewer: Doing so should make it look like this:

> [!NOTE]
> Think about the maintenance cost of adding a new E2E test. Always favour an integration test if possible.

> For new features, we should decide as part of the planning or shaping of work/tickets whether an e2e test is necessary for the feature.

Reviewer: ```suggestion
For new features, we should decide as part of the planning or shaping of work/tickets whether an E2E test is necessary for the feature.
```

### What should we automate?

> The decision to write an e2e test should be taken during planning as a team, but these are general guidelines to what should go into e2e tests

Reviewer: ```suggestion
The decision to write an E2E test should be taken during planning as a team, but these are general guidelines to what should go into E2E tests
```

#### Critical paths 
The critical path is the set of components and workflows that is required for the application to serve its core function. For example, an article page displaying the content and title for the given service is critical in our context, but links to related topics or top stories, can be considered non-critical. These should be tested as well, but they’re better candidates for a less expensive layer of testing like integration tests.

#### Smoke tests
We should have smoke tests for validating that core functionality is not broken, for example, that the page loads, has titles and content. These should run as often as possible (on each PR), they should be quick, reliable and repeatable (not flaky).

#### As little as possible
The rule of thumb is that we should avoid writing an end to end test if there is an alternative to validating the functionality. The alternative can be a combination of layers, for example, an integration test combined with an API test (which doesn’t exist right now) can be enough to reach the same confidence as an e2e test with less cost.

### Puma approach

> Adopting the [PUMA](https://confluence.dev.bbc.co.uk/display/podtest/PUMA) approach can provide a framework to decide if a proposed test genuinely meets the definition of an E2E. According to PUMA, an e2e test should:

Reviewer: ```suggestion
Adopting the [PUMA](https://confluence.dev.bbc.co.uk/display/podtest/PUMA) approach can provide a framework to decide if a proposed test genuinely meets the definition of an E2E. According to PUMA, an E2E test should:
```

> P - Prove core functionality, 
> U - be Understood by all,
> M - Mandatory
> A - Automated

Reviewer: We had to add extra lines to make them appear as we want - this is what it looks like: ![Screenshot 2024-11-28 at 12 49 24](https://github.com/user-attachments/assets/7d66de41-0e64-4a63-8966-7ea97a6f96a1)

```suggestion
P - Prove core functionality

U - be Understood by all

M - Mandatory

A - Automated
```

### How should we automate?

> Patterns, anti-patterns and practices for e2e tests

Reviewer: ```suggestion
Patterns, anti-patterns and practices for E2E tests
```

> In this section, we will document some of the patterns we use and the best practices to write better e2e tests.

Reviewer: ```suggestion
In this section, we will document some of the patterns we use and the best practices to write better E2E tests.
```

#### Cypress best practices
A good place to start is the Cypress best practices page: https://docs.cypress.io/guides/references/best-practices

> Custom commands for common functionalities

Reviewer: Does this need to be a heading?

> We have many custom commands that make common tasks easier, these are located under [cypress/e2e/commands](https://github.com/bbc/simorgh/blob/latest/cypress/support/commands/application.js). It’s good to have a look at these to have an overview of what’s possible, and consider abstracting common functionality as a [cypress command](https://docs.cypress.io/api/cypress-api/custom-commands) when it makes sense.

Reviewer: ```suggestion
We have many custom commands that make common tasks easier, these are located within [cypress/e2e/commands](https://github.com/bbc/simorgh/blob/latest/cypress/support/commands/application.js). It’s good to have a look at these to have an overview of what’s possible, and consider abstracting common functionality as a [cypress command](https://docs.cypress.io/api/cypress-api/custom-commands) when it makes sense.
```

> 
> If the test is “green” (i.e. it confirms the actual behaviour matches the expected behaviour), then we are provided with the knowledge that at that time the system behaved as defined by the test.
> 
> If the test is “red” (i.e. it confirms the actual behaviour did not match expectation), then we are provided with the knowledge that at that time the system did not behave as defined by the test.
> 
> If a test remains permanently “red”, then we either misunderstood the purpose of the code and so need to review the expectations of the checks in the test or we have uncovered unexpected behaviour in the system under test.
> 
> If a test remains permanently “green”, then we confirm current behaviour continues to meet the expectations of the test.
> 
> However, a test that flickers between the two states indicates something is not right and should not be ignored! I believe describing these as flaky makes it easy to ignore potential problems either in the functionality of the system under test or the test itself.
> 

> > quote from  @Simon F 

Reviewer: Let's remove this, or else, attribute the quote to Simon's github username instead?

> 🔊 Discussion with team: What are other good practices? any anti-patterns?

Who should automate?

> Responsibilities

Reviewer: Should this be a heading?

|             | Unit tests         | Integration        | E2E                |
| ----------- | ------------------ | ------------------ | ------------------ |
| Responsible | Developers         | Developers and QA* | QA and Developers* |
| Accountable | Developers         | Developers         | QA                 |
| Consulted** | Developers and QA# | Developers and QA  | Developers and QA  |
| Informed**  | Developers and QA# | Developers and QA  | Developers and QA  |

> * For integration tests, developers are accountable for them, but sometimes QA can be involved in their implementation. For e2e, QA are accountable for them but developers can be involved in the implementation too.

Reviewer: ```suggestion
* For integration tests, developers are accountable for them, but sometimes QA can be involved in their implementation. For E2E, QA are accountable for them but developers can be involved in the implementation too.
```

** Consulting and informing in this context happens mainly during the sprint planning or three amigos. As a team, we can decide on whether a feature requires an integration or e2e test (or both). Consulting also includes code reviews of tests being added.

> ##### There is no reason why QA cannot be consulted and informed about the content of unit tests. As specialists, they might consider and uncover edge cases/scenarios that do not occur to developers and can also help to review the language used to frame the tests (think about `describe` and `it` blocks) even if they are not confident reading the test code itself.

Reviewer: I don't think this should be a header - but rather a note / quote / bold?

Would this work?

```suggestion

> [!NOTE] 
> There is no reason why QA cannot be consulted and informed about the content of unit tests. As specialists, they might consider and uncover edge cases/scenarios that do not occur to developers and can also help to review the language used to frame the tests (think about `describe` and `it` blocks) even if they are not confident reading the test code itself.
```

Reviewer: Doing so will make it appear like this:

> [!NOTE] 
> There is no reason why QA cannot be consulted and informed about the content of unit tests. As specialists, they might consider and uncover edge cases/scenarios that do not occur to developers and can also help to review the language used to frame the tests (think about `describe` and `it` blocks) even if they are not confident reading the test code itself.

> ##### There is no reason why QA cannot be consulted and informed about the content of unit tests. As specialists, they might consider and uncover edge cases/scenarios that do not occur to developers and can also help to review the language used to frame the tests (think about `describe` and `it` blocks) even if they are not confident reading the test code itself.

Reviewer: Should this be writing instead of reading?

```suggestion
##### There is no reason why QA cannot be consulted and informed about the content of unit tests. As specialists, they might consider and uncover edge cases/scenarios that do not occur to developers and can also help to review the language used to frame the tests (think about `describe` and `it` blocks) even if they are not confident writing the test code itself.
```

##### References
- [Five Factor Testing](https://madeintandem.com/blog/five-factor-testing/)
- [Practical test pyramid](https://martinfowler.com/articles/practical-test-pyramid.html)

> - https://confluence.dev.bbc.co.uk/display/podtest/PUMA

Reviewer: Should we be referencing an internal confluence page, or would it be better to find a public one?

5. Support refactoring

[source: https://madeintandem.com/blog/five-factor-testing/]

> 🔊 Discussion point with team:
> - Any other reasons? 
> - What are for you the most important reasons?
    

It is important to think about why we write tests because it influences what types of testing we should prioritise. For example, unit tests might be good for improving the system design as the units become more independent and easier to test, but having only unit tests doesn’t necessarily verify the code is working correctly (beyond the single unit) as it won’t necessarily reveal problems that occur in the integration of components. On the other hand, E2E tests can provide better verification of the software, but they’re expensive and would not necessarily improve software design, nor document the code behaviour at unit or component level.

## Where should we automate?
### Defining the test levels

Broadly speaking, we follow the test pyramid. We have a large set of unit tests, a smaller subset of integration tests, and an even smaller subset of UI tests. We should continue to follow this pattern.

![Image](https://blog.getmason.io/content/images/2020/11/Testing-pyramid--6--1.jpg)

> Image [Source](https://blog.getmason.io/content/images/2020/11/Testing-pyramid--6--1.jpg)
> To avoid confusion, we will clarify what the test levels mean in our context.

Reviewer: Add whitespace to put text on new line

```suggestion
Image [Source](https://blog.getmason.io/content/images/2020/11/Testing-pyramid--6--1.jpg)

To avoid confusion, we will clarify what the test levels mean in our context.
```

### Unit tests
We use jest to write unit tests. They normally exercise a single component (UI components, hooks, helper methods). Here is an [example unit test](https://github.com/bbc/simorgh/tree/bfbed54d2b1fe880e8f171023054c16d1c274ca2/src/app/components/Heading).
Some patterns:

- We use [data-driven tests](https://github.com/bbc/simorgh/blob/bfbed54d2b1fe880e8f171023054c16d1c274ca2/src/app/hooks/useImageColour/index.test.js#L10) to define inputs to tests

> - We use [snapshot tests](https://github.com/bbc/simorgh/blob/4f385d94bee96af48eff2eca49c24cf382a3a494/src/app/components/FrostedGlassPromo/index.test.tsx#L76) to easily test certain scenarios such as (TODO what does this test show us?)

Reviewer: What does this test show us - we need to remove the TODO

> [!NOTE]
 Think about the maintenance cost of adding a new E2E test. Always favour an integration test if possible instead of an end to end test.

For new features, we should decide as part of the planning or shaping of work/tickets whether an E2E test is necessary for the feature.

### What should we automate?
The decision to write an E2E test should be taken during planning as a team, but these are general guidelines to what should go into E2E tests

#### Critical paths 
The critical path is the set of components and workflows that is required for the application to serve its core function. For example, an article page displaying the content and title for the given service is critical in our context, but links to related topics or top stories, can be considered non-critical. These should be tested as well, but they’re better candidates for a less expensive layer of testing like integration tests.

#### Smoke tests
We should have smoke tests for validating that core functionality is not broken, for example, that the page loads, has titles and content. These should run as often as possible (on each PR), they should be quick, reliable and repeatable (not flaky).

> #### As little as possible
> The rule of thumb is that we should avoid writing an end to end test if there is an alternative to validating the functionality. The alternative can be a combination of layers, for example, an integration test combined with an API test (which doesn’t exist right now) can be enough to reach the same confidence as an e2e test with less cost.

Reviewer: ```suggestion
#### As little as possible
The rule of thumb is that we should avoid writing an end to end test if there is an alternative to validating the functionality. The alternative can be a combination of layers, for example, an integration test combined with an API test (which doesn’t exist right now) can be enough to reach the same confidence as an E2E test with less cost.
```

> 
> If the test is “green” (i.e. it confirms the actual behaviour matches the expected behaviour), then we are provided with the knowledge that at that time the system behaved as defined by the test.
> 
> If the test is “red” (i.e. it confirms the actual behaviour did not match expectation), then we are provided with the knowledge that at that time the system did not behave as defined by the test.
> 
> If a test remains permanently “red”, then we either misunderstood the purpose of the code and so need to review the expectations of the checks in the test or we have uncovered unexpected behaviour in the system under test.
> 
> If a test remains permanently “green”, then we confirm current behaviour continues to meet the expectations of the test.
> 
> However, a test that flickers between the two states indicates something is not right and should not be ignored! I believe describing these as flaky makes it easy to ignore potential problems either in the functionality of the system under test or the test itself.
> 

> > quote from  @sframpton

Reviewer: This is not rendering as expected. I think we need to remove it from the > section
```suggestion
[Simon Frampton, Principal Test Engineer, BBC](@sframpton)
```

> > 🔊 Discussion with team: 
> > What are other good practices?
> > Any anti-patterns?

Reviewer: ```suggestion
> 🔊 Discussion with team: 
> - What are other good practices?
> - Any anti-patterns?
```

#### Who should automate?
#### Responsibilities
|             | Unit tests         | Integration        | E2E                |
| ----------- | ------------------ | ------------------ | ------------------ |
| Responsible | Developers         | Developers and QA* | QA and Developers* |
| Accountable | Developers         | Developers         | QA                 |
| Consulted** | Developers and QA# | Developers and QA  | Developers and QA  |
| Informed**  | Developers and QA# | Developers and QA  | Developers and QA  |

* For integration tests, developers are accountable for them, but sometimes QA can be involved in their implementation. For E2E, QA are accountable for them but developers can be involved in the implementation too.

> ** Consulting and informing in this context happens mainly during the sprint planning or three amigos. As a team, we can decide on whether a feature requires an integration or e2e test (or both). Consulting also includes code reviews of tests being added.

Reviewer: ```suggestion
** Consulting and informing in this context happens mainly during the sprint planning or three amigos. As a team, we can decide on whether a feature requires an integration or E2E test (or both). Consulting also includes code reviews of tests being added.
```

> |             | Unit tests         | Integration        | E2E                |
> | ----------- | ------------------ | ------------------ | ------------------ |
> | Responsible | Developers         | Developers and QA* | QA and Developers* |
> | Accountable | Developers         | Developers         | QA                 |
> | Consulted** | Developers and QA# | Developers and QA  | Developers and QA  |
> | Informed**  | Developers and QA# | Developers and QA  | Developers and QA  |
> 
> * For integration tests, developers are accountable for them, but sometimes QA can be involved in their implementation. For E2E, QA are accountable for them but developers can be involved in the implementation too.
> 
> ** Consulting and informing in this context happens mainly during the sprint planning or three amigos. As a team, we can decide on whether a feature requires an integration or e2e test (or both). Consulting also includes code reviews of tests being added.

Reviewer: These points are not rendering as expected. Could we use a different symbol (not * as it's used in markdown to make text bold) 

```suggestion
|             | Unit tests         | Integration        | E2E                |
| ----------- | ------------------ | ------------------ | ------------------ |
| Responsible | Developers         | Developers and QA (^) | QA and Developers (^) |
| Accountable | Developers         | Developers         | QA                 |
| Consulted (+) | Developers and QA# | Developers and QA  | Developers and QA  |
| Informed (+)  | Developers and QA# | Developers and QA  | Developers and QA  |

(^) For integration tests, developers are accountable for them, but sometimes QA can be involved in their implementation. For E2E, QA are accountable for them but developers can be involved in the implementation too.

(+) Consulting and informing in this context happens mainly during the sprint planning or three amigos. As a team, we can decide on whether a feature requires an integration or e2e test (or both). Consulting also includes code reviews of tests being added.
```
